In [1]:
%matplotlib inline
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
import numpy as np
import pandas as pd

In [2]:
def area(box):
    return (box[2] - box[0]) * (box[3] - box[1])


def intersection_over_union(boxes):
    assert(len(boxes) == 8)
    boxA = boxes[:4].values
    boxB = boxes[4:].values
    
    boxAArea = area(boxA)
    boxBArea = area(boxB)
    
    if (boxAArea == 0 or boxBArea == 0):
        return 0
        
    xA = max(boxA[0], boxB[0])
    yA = max(boxA[1], boxB[1])
    xB = min(boxA[2], boxB[2])
    yB = min(boxA[3], boxB[3])

    interArea = max(0, xB - xA) * max(0, yB - yA)

    
    iou = interArea / float(boxAArea + boxBArea - interArea)
    return iou

## Импорт данных

In [3]:
votes = pd.read_csv("train_data.csv")

In [4]:
answers = pd.read_csv("train_answers.csv")

## Фильтрация выбросов по интерквартильному размаху

In [56]:
def iqr(qr1, qr3):
    return (qr3 - qr1) * 1.5

In [64]:
def calc_iqrs(votez):
    q1_quantiles = votez.groupby('itemId')[['Xmin', 'Ymin', 'Xmax', 'Ymax']].quantile(q=[0.25]).reset_index().drop(columns=['level_1'])
    q3_quantiles = votez.groupby('itemId')[['Xmin', 'Ymin', 'Xmax', 'Ymax']].quantile(q=[0.75]).reset_index().drop(columns=['level_1'])
    q_merged = q1_quantiles.merge(q3_quantiles, on='itemId', suffixes=('_q1', '_q3'))
    q_merged['Xmin_iqr'] = iqr(q_merged['Xmin_q1'], q_merged['Xmin_q3'])
    q_merged['Ymin_iqr'] = iqr(q_merged['Ymin_q1'], q_merged['Ymin_q3'])
    q_merged['Xmax_iqr'] = iqr(q_merged['Xmax_q1'], q_merged['Xmax_q3'])
    q_merged['Ymax_iqr'] = iqr(q_merged['Ymax_q1'], q_merged['Ymax_q3'])
    q_merged = q_merged.drop(columns=['Xmin_q1', 'Ymin_q1', 'Xmax_q1', 'Ymax_q1', 'Xmin_q3', 'Ymin_q3', 'Xmax_q3', 'Ymax_q3'])
    return q_merged

In [69]:
def calc_low_high_iqrs(votez, iqrs):
    q2_quantiles = votez.groupby('itemId')[['Xmin', 'Ymin', 'Xmax', 'Ymax']].median().reset_index()
    q_filter = iqrs.merge(q2_quantiles, on='itemId')
    q_filter['Xmin_low'] = q_filter['Xmin'] - q_filter['Xmin_iqr']
    q_filter['Ymin_low'] = q_filter['Ymin'] - q_filter['Ymin_iqr']
    q_filter['Xmax_low'] = q_filter['Xmax'] - q_filter['Xmax_iqr']
    q_filter['Ymax_low'] = q_filter['Ymax'] - q_filter['Ymax_iqr']
    q_filter['Xmin_high'] = q_filter['Xmin'] + q_filter['Xmin_iqr']
    q_filter['Ymin_high'] = q_filter['Ymin'] + q_filter['Ymin_iqr']
    q_filter['Xmax_high'] = q_filter['Xmax'] + q_filter['Xmax_iqr']
    q_filter['Ymax_high'] = q_filter['Ymax'] + q_filter['Ymax_iqr']
    q_filter = q_filter.drop(columns=['Xmin', 'Ymin', 'Xmax', 'Ymax', 'Xmin_iqr', 'Ymin_iqr', 'Xmax_iqr', 'Ymax_iqr'])
    return q_filter

In [91]:
def filter_by_iqr(votez):
    print('изображений до фильтрации:', len(votez['itemId'].unique()))
    q_merged = calc_iqrs(votez)
    q_filter = calc_low_high_iqrs(votez, q_merged)
    filter_func = lambda df: (
        (df['Xmin'] >= df['Xmin_low']) & (df['Xmin'] <= df['Xmin_high']) &
        (df['Ymin'] >= df['Ymin_low']) & (df['Ymin'] <= df['Ymin_high']) &
        (df['Xmax'] >= df['Xmax_low']) & (df['Xmax'] <= df['Xmax_high']) &
        (df['Ymax'] >= df['Ymax_low']) & (df['Ymax'] <= df['Ymax_high'])
    )
    # изменить логику. Если значение выходит за рамки интерквартильного интервала, то присваивать ему среднее значение
    # сейчас часть данных отсеилось за счёт выбросов хотя бы в одном из 4 входных свойств
    votes_filtered = votez.merge(q_filter, on='itemId')#.loc[lambda df: filter_func(df), :].drop(columns=['Xmin_low', 'Ymin_low', 'Xmax_low', 'Ymax_low', 'Xmin_high', 'Ymin_high', 'Xmax_high', 'Ymax_high'])
    print('изображений после фильтрации:', len(votes_filtered['itemId'].unique()))
    return votes_filtered

In [95]:
# set(votes['itemId'].unique()) - set(filter_by_iqr(votes)['itemId'].unique())
filter_by_iqr(votes) \
.loc[lambda df: df['itemId'].isin({678, 7761, 8398, 8774, 10521, 15247, 16237, 21836, 22723, 30034, 34214, 34339}), :] \
.sort_values('itemId')

изображений до фильтрации: 943
изображений после фильтрации: 943


userId  itemId  Xmin  Ymin  Xmax  Ymax  Xmin_low  Ymin_low  Xmax_low  \
1388     366     678    63   593   213   738     39.00    580.75    195.00   
1387     141     678    32   586   210   739     39.00    580.75    195.00   
1386     903     678    64   586   193   712     39.00    580.75    195.00   
2587    1508    7761    22   452    31   475     21.25    451.25     30.75   
2588    1389    7761    23   452    42   476     21.25    451.25     30.75   
2586    1442    7761    22   453    46   476     21.25    451.25     30.75   
12        29    8398    53  1020    61  1027     52.25   1019.25     61.00   
14      1050    8398    53  1019    61  1028     52.25   1019.25     61.00   
13       421    8398    54  1020    61  1028     52.25   1019.25     61.00   
17       311    8774   288   967   296   975    144.00    903.25    380.75   
16       866    8774   308  1044   755  1317    144.00    903.25    380.75   
15       154    8774   116   959   795  1300    144.00    903.25    380.75   
189     1599   10521   100   868   315   994     56.50    827.50    313.50   
190      886   10521    55   835   318   995     56.50    827.50    313.50   
191      392   10521    66   850   314   999     56.50    827.50    313.50   
192     1696   10521    67   837   310   994     56.50    827.50    313.50   
188      283   10521    73   847   315   992     56.50    827.50    313.50   
2710    1432   15247    78   261   827   590     76.25     10.00    731.00   
2709    1328   15247   137   121   727   587     76.25     10.00    731.00   
2711    1657   15247   159   113   806   601     76.25     10.00    731.00   
3609    1107   16237    34   491   429   735      7.00    484.25    353.75   
3610      58   16237    74   488   413   677      7.00    484.25    353.75   
3611     322   16237    37   497   350   684      7.00    484.25    353.75   
3716     851   21836   205   349   603   697    208.50    311.75    542.50   
3714    1182   21836   222   400   586   703    208.50    311.75    542.50   
3715     853   21836   223   350   545   729    208.50    311.75    542.50   
4230     656   22723    95   416   371   654     86.50    429.00    362.00   
4231    1254   22723    85   476   371   637     86.50    429.00    362.00   
4232     539   22723    94   474   359   592     86.50    429.00    362.00   
5090     119   30034    89   679   640   970     46.75    555.50    648.25   
5089    1370   30034    82   605   673  1026     46.75    555.50    648.25   
5088     886   30034    42   611   673   973     46.75    555.50    648.25   
5094      32   34214    54   371   849   998     -8.00    332.00    838.25   
5096     759   34214    14   331   848  1017     -8.00    332.00    838.25   
5095    1384   34214    22   362   836  1002     -8.00    332.00    838.25   
1640    1304   34339   107   552   426   896     45.25    553.00    307.75   
1641    1155   34339   106   572   370   843     45.25    553.00    307.75   
1642    1097   34339    26   568   343   886     45.25    553.00    307.75   

      Ymax_low  Xmin_high  Ymin_high  Xmax_high  Ymax_high  
1388    717.75      87.00     591.25     225.00     758.25  
1387    717.75      87.00     591.25     225.00     758.25  
1386    717.75      87.00     591.25     225.00     758.25  
2587    475.25      22.75     452.75      53.25     476.75  
2588    475.25      22.75     452.75      53.25     476.75  
2586    475.25      22.75     452.75      53.25     476.75  
12     1027.25      53.75    1020.75      61.00    1028.75  
14     1027.25      53.75    1020.75      61.00    1028.75  
13     1027.25      53.75    1020.75      61.00    1028.75  
17     1043.50     432.00    1030.75    1129.25    1556.50  
16     1043.50     432.00    1030.75    1129.25    1556.50  
15     1043.50     432.00    1030.75    1129.25    1556.50  
189     992.50      77.50     866.50     316.50     995.50  
190     992.50      77.50     866.50     316.50     995.50  
191     992.50      77.50     866.50     316.50 

## Расчёт метрики Intersection over Union (IoU)

In [23]:
quorum = votes.groupby("itemId")[['Xmin','Ymin', 'Xmax', 'Ymax']].mean().reset_index()

In [24]:
data = quorum.merge(answers, on=["itemId"])

In [25]:
data["iou"] = data[['Xmin','Ymin', 'Xmax', 'Ymax', 'Xmin_true',\
      'Ymin_true', 'Xmax_true','Ymax_true']].apply(intersection_over_union, axis=1)

In [26]:
data["iou"].mean()

0.5058725791429202

### История изменений
* 0.5058725791429202 – начальное значение
* 0.5129929061634199 – фильтрация по интерквартильному размаху

# Экспорт тестовых данных

In [82]:
votes_test = pd.read_csv("test_data.csv")
votes_quorum = filter_by_iqr(votes_test).groupby("itemId")[['Xmin','Ymin', 'Xmax', 'Ymax']].mean().reset_index()
# votes_quorum.columns = ['itemId', 'Xmin_true', 'Ymin_true', 'Xmax_true','Ymax_true']
votes_quorum.round().astype(int).to_csv('AIatsuk_predictions.csv', index=False)

изображений до фильтрации: 630
изображений после фильтрации: 622
